### PR TITLE
executor: fix DATA RACE for reading information_schema.columns where there are views

### LINKED ISSUE #34509: DATA RACE in tryFillViewColumnType.func1()
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

Found in CI race test https://ci.pingcap.net/blue/organizations/jenkins/ut-check/detail/ut-check/10630/pipeline/

### 2. What did you expect to see? (Required)

No DATA RACE

### 3. What did you see instead (Required)

```
[2022-05-09T16:30:59.876Z] ==================
[2022-05-09T16:30:59.876Z] WARNING: DATA RACE
[2022-05-09T16:30:59.876Z] Read at 0x00c00472c210 by goroutine 106:
[2022-05-09T16:30:59.876Z]   github.com/pingcap/tidb/session.(*session).SetProcessInfo()
[2022-05-09T16:30:59.876Z]       /home/jenkins/agent/workspace/ut-check/tidb/session/session.go:1390 +0x44a
[2022-05-09T16:30:59.876Z]   github.com/pingcap/tidb/session.(*session).ExecuteStmt()
[2022-05-09T16:30:59.876Z]       /home/jenkins/agent/workspace/ut-check/tidb/session/session.go:1893 +0x592
[2022-05-09T16:30:59.876Z]   github.com/pingcap/tidb/session.(*session).ExecRestrictedSQL.func1()
[2022-05-09T16:30:59.876Z]       /home/jenkins/agent/workspace/ut-check/tidb/session/session.go:1838 +0x2b7
[2022-05-09T16:30:59.876Z]   github.com/pingcap/tidb/session.(*session).withRestrictedSQLExecutor()
[2022-05-09T16:30:59.876Z]       /home/jenkins/agent/workspace/ut-check/tidb/session/session.go:1824 +0x391
[2022-05-09T16:30:59.876Z]   github.com/pingcap/tidb/session.(*session).ExecRestrictedSQL()
[2022-05-09T16:30:59.876Z]       /home/jenkins/agent/workspace/ut-check/tidb/session/session.go:1828 +0x124
[2022-05-09T16:30:59.876Z]   github.com/pingcap/tidb/bindinfo.(*BindHandle).Update()
[2022-05-09T16:30:59.876Z]       /home/jenkins/agent/workspace/ut-check/tidb/bindinfo/handle.go:139 +0x1e9
[2022-05-09T16:30:59.876Z]   github.com/pingcap/tidb/domain.(*Domain).globalBindHandleWorkerLoop.func1()
[2022-05-09T16:30:59.876Z]       /home/jenkins/agent/workspace/ut-check/tidb/domain/domain.go:1137 +0x2be
[2022-05-09T16:30:59.876Z] 
[2022-05-09T16:30:59.876Z] Previous write at 0x00c00472c210 by goroutine 66:
[2022-05-09T16:30:59.876Z]   github.com/pingcap/tidb/executor.runWithSystemSession.func1()
[2022-05-09T16:30:59.876Z]       /home/jenkins/agent/workspace/ut-check/tidb/executor/show.go:1937 +0x84
[2022-05-09T16:30:59.876Z]   runtime.deferreturn()
[2022-05-09T16:30:59.876Z]       /usr/local/go/src/runtime/panic.go:436 +0x32
[2022-05-09T16:30:59.876Z]   github.com/pingcap/tidb/executor.tryFillViewColumnType()
[2022-05-09T16:30:59.876Z]       /home/jenkins/agent/workspace/ut-check/tidb/executor/show.go:1904 +0x16d
[2022-05-09T16:30:59.876Z]   github.com/pingcap/tidb/executor.(*hugeMemTableRetriever).dataForColumnsInTable()
[2022-05-09T16:30:59.876Z]       /home/jenkins/agent/workspace/ut-check/tidb/executor/infoschema_reader.go:680 +0x159
[2022-05-09T16:30:59.876Z]   github.com/pingcap/tidb/executor.(*hugeMemTableRetriever).setDataForColumns()
[2022-05-09T16:30:59.876Z]       /home/jenkins/agent/workspace/ut-check/tidb/executor/infoschema_reader.go:669 +0x4a4
[2022-05-09T16:30:59.876Z]   github.com/pingcap/tidb/executor.(*hugeMemTableRetriever).retrieve()
[2022-05-09T16:30:59.876Z]       /home/jenkins/agent/workspace/ut-check/tidb/executor/infoschema_reader.go:2614 +0x314
[2022-05-09T16:30:59.876Z]   github.com/pingcap/tidb/executor.(*MemTableReaderExec).Next()
[2022-05-09T16:30:59.876Z]       /home/jenkins/agent/workspace/ut-check/tidb/executor/memtable_reader.go:118 +0x482
[2022-05-09T16:30:59.876Z]   github.com/pingcap/tidb/executor.Next()
[2022-05-09T16:30:59.876Z]       /home/jenkins/agent/workspace/ut-check/tidb/executor/executor.go:306 +0x645
[2022-05-09T16:30:59.876Z]   github.com/pingcap/tidb/executor.(*SelectionExec).Next()
[2022-05-09T16:30:59.876Z]       /home/jenkins/agent/workspace/ut-check/tidb/executor/executor.go:1398 +0x278
[2022-05-09T16:30:59.876Z]   github.com/pingcap/tidb/executor.Next()
[2022-05-09T16:30:59.876Z]       /home/jenkins/agent/workspace/ut-check/tidb/executor/executor.go:306 +0x645
[2022-05-09T16:30:59.876Z]   github.com/pingcap/tidb/executor.(*HashJoinExec).fetchBuildSideRows()
[2022-05-09T16:30:59.876Z]       /home/jenkins/agent/workspace/ut-check/tidb/executor/join.go:282 +0x43e
[2022-05-09T16:30:59.876Z]   github.com/pingcap/tidb/executor.(*HashJoinExec).fetchAndBuildHashTable.func2()
[2022-05-09T16:30:59.876Z]       /home/jenkins/agent/workspace/ut-check/tidb/executor/join.go:731 +0xf9
[2022-05-09T16:30:59.876Z]   github.com/pingcap/tidb/util.WithRecovery()
[2022-05-09T16:30:59.876Z]       /home/jenkins/agent/workspace/ut-check/tidb/util/misc.go:90 +0x6d
[2022-05-09T16:30:59.876Z]   github.com/pingcap/tidb/executor.(*HashJoinExec).fetchAndBuildHashTable.func4()
[2022-05-09T16:30:59.876Z]       /home/jenkins/agent/workspace/ut-check/tidb/executor/join.go:728 +0x47
[2022-05-09T16:30:59.876Z] 
[2022-05-09T16:30:59.876Z] Goroutine 106 (running) created at:
[2022-05-09T16:30:59.876Z]   github.com/pingcap/tidb/domain.(*Domain).globalBindHandleWorkerLoop()
[2022-05-09T16:30:59.876Z]       /home/jenkins/agent/workspace/ut-check/tidb/domain/domain.go:1120 +0xef
[2022-05-09T16:30:59.876Z]   github.com/pingcap/tidb/domain.(*Domain).LoadBindInfoLoop()
[2022-05-09T16:30:59.876Z]       /home/jenkins/agent/workspace/ut-check/tidb/domain/domain.go:1113 +0x1a4
[2022-05-09T16:30:59.876Z]   github.com/pingcap/tidb/session.BootstrapSession()
[2022-05-09T16:30:59.876Z]       /home/jenkins/agent/workspace/ut-check/tidb/session/session.go:2876 +0x4c5
[2022-05-09T16:30:59.876Z]   github.com/pingcap/tidb/testkit.bootstrap()
[2022-05-09T16:30:59.876Z]       /home/jenkins/agent/workspace/ut-check/tidb/testkit/mockstore.go:49 +0x8c
[2022-05-09T16:30:59.876Z]   github.com/pingcap/tidb/testkit.CreateMockStoreAndDomain()
[2022-05-09T16:30:59.876Z]       /home/jenkins/agent/workspace/ut-check/tidb/testkit/mockstore.go:42 +0xcd
[2022-05-09T16:30:59.876Z]   github.com/pingcap/tidb/testkit.CreateMockStore()
[2022-05-09T16:30:59.876Z]       /home/jenkins/agent/workspace/ut-check/tidb/testkit/mockstore.go:34 +0x55
[2022-05-09T16:30:59.876Z]   github.com/pingcap/tidb/executor_test.TestJoinSystemTableContainsView()
[2022-05-09T16:30:59.876Z]       /home/jenkins/agent/workspace/ut-check/tidb/executor/infoschema_reader_test.go:644 +0x26
[2022-05-09T16:30:59.876Z]   testing.tRunner()
[2022-05-09T16:30:59.876Z]       /usr/local/go/src/testing/testing.go:1439 +0x213
[2022-05-09T16:30:59.876Z]   testing.(*T).Run.func1()
[2022-05-09T16:30:59.876Z]       /usr/local/go/src/testing/testing.go:1486 +0x47

[2022-05-09T16:30:59.876Z] ==================
[2022-05-09T16:30:59.876Z] ==================
[2022-05-09T16:30:59.876Z] WARNING: DATA RACE
[2022-05-09T16:30:59.876Z] Write at 0x00c0050eb868 by goroutine 68:
[2022-05-09T16:30:59.876Z]   github.com/pingcap/tidb/executor.tryFillViewColumnType.func1()
[2022-05-09T16:30:59.876Z]       /home/jenkins/agent/workspace/ut-check/tidb/executor/show.go:1913 +0x645
[2022-05-09T16:30:59.876Z]   github.com/pingcap/tidb/executor.runWithSystemSession()
[2022-05-09T16:30:59.876Z]       /home/jenkins/agent/workspace/ut-check/tidb/executor/show.go:1940 +0x334
[2022-05-09T16:30:59.877Z]   github.com/pingcap/tidb/executor.tryFillViewColumnType()
[2022-05-09T16:30:59.877Z]       /home/jenkins/agent/workspace/ut-check/tidb/executor/show.go:1904 +0x16d
[2022-05-09T16:30:59.877Z]   github.com/pingcap/tidb/executor.(*hugeMemTableRetriever).dataForColumnsInTable()
[2022-05-09T16:30:59.877Z]       /home/jenkins/agent/workspace/ut-check/tidb/executor/infoschema_reader.go:680 +0x159
[2022-05-09T16:30:59.877Z]   github.com/pingcap/tidb/executor.(*hugeMemTableRetriever).setDataForColumns()
[2022-05-09T16:30:59.877Z]       /home/jenkins/agent/workspace/ut-check/tidb/executor/infoschema_reader.go:669 +0x4a4
[2022-05-09T16:30:59.877Z]   github.com/pingcap/tidb/executor.(*hugeMemTableRetriever).retrieve()
[2022-05-09T16:30:59.877Z]       /home/jenkins/agent/workspace/ut-check/tidb/executor/infoschema_reader.go:2614 +0x314
[2022-05-09T16:30:59.877Z]   github.com/pingcap/tidb/executor.(*MemTableReaderExec).Next()
[2022-05-09T16:30:59.877Z]       /home/jenkins/agent/workspace/ut-check/tidb/executor/memtable_reader.go:118 +0x482
[2022-05-09T16:30:59.877Z]   github.com/pingcap/tidb/executor.Next()
[2022-05-09T16:30:59.877Z]       /home/jenkins/agent/workspace/ut-check/tidb/executor/executor.go:306 +0x645
[2022-05-09T16:30:59.877Z]   github.com/pingcap/tidb/executor.(*SelectionExec).Next()
[2022-05-09T16:30:59.877Z]       /home/jenkins/agent/workspace/ut-check/tidb/executor/executor.go:1398 +0x278
[2022-05-09T16:30:59.877Z]   github.com/pingcap/tidb/executor.Next()
[2022-05-09T16:30:59.877Z]       /home/jenkins/agent/workspace/ut-check/tidb/executor/executor.go:306 +0x645
[2022-05-09T16:30:59.877Z]   github.com/pingcap/tidb/executor.(*HashJoinExec).fetchBuildSideRows()
[2022-05-09T16:30:59.877Z]       /home/jenkins/agent/workspace/ut-check/tidb/executor/join.go:282 +0x43e
[2022-05-09T16:30:59.877Z]   github.com/pingcap/tidb/executor.(*HashJoinExec).fetchAndBuildHashTable.func2()
[2022-05-09T16:30:59.877Z]       /home/jenkins/agent/workspace/ut-check/tidb/executor/join.go:731 +0xf9
[2022-05-09T16:30:59.877Z]   github.com/pingcap/tidb/util.WithRecovery()
[2022-05-09T16:30:59.877Z]       /home/jenkins/agent/workspace/ut-check/tidb/util/misc.go:90 +0x6d
[2022-05-09T16:30:59.877Z]   github.com/pingcap/tidb/executor.(*HashJoinExec).fetchAndBuildHashTable.func4()
[2022-05-09T16:30:59.877Z]       /home/jenkins/agent/workspace/ut-check/tidb/executor/join.go:728 +0x47
[2022-05-09T16:30:59.877Z] 
[2022-05-09T16:30:59.877Z] Previous read at 0x00c0050eb868 by goroutine 66:
[2022-05-09T16:30:59.877Z]   github.com/pingcap/tidb/parser/types.(*FieldType).GetType()
[2022-05-09T16:30:59.877Z]       /home/jenkins/agent/workspace/ut-check/tidb/parser/types/field_type.go:63 +0x11f4
[2022-05-09T16:30:59.877Z]   github.com/pingcap/tidb/parser/model.(*ColumnInfo).GetType()
[2022-05-09T16:30:59.877Z]       /home/jenkins/agent/workspace/ut-check/tidb/parser/model/model.go:135 +0x11e7
[2022-05-09T16:30:59.877Z]   github.com/pingcap/tidb/executor.(*hugeMemTableRetriever).dataForColumnsInTable()
[2022-05-09T16:30:59.877Z]       /home/jenkins/agent/workspace/ut-check/tidb/executor/infoschema_reader.go:811 +0xd84
[2022-05-09T16:30:59.877Z]   github.com/pingcap/tidb/executor.(*hugeMemTableRetriever).setDataForColumns()
[2022-05-09T16:30:59.877Z]       /home/jenkins/agent/workspace/ut-check/tidb/executor/infoschema_reader.go:669 +0x4a4
[2022-05-09T16:30:59.877Z]   github.com/pingcap/tidb/executor.(*hugeMemTableRetriever).retrieve()
[2022-05-09T16:30:59.877Z]       /home/jenkins/agent/workspace/ut-check/tidb/executor/infoschema_reader.go:2614 +0x314
[2022-05-09T16:30:59.877Z]   github.com/pingcap/tidb/executor.(*MemTableReaderExec).Next()
[2022-05-09T16:30:59.877Z]       /home/jenkins/agent/workspace/ut-check/tidb/executor/memtable_reader.go:118 +0x482
[2022-05-09T16:30:59.877Z]   github.com/pingcap/tidb/executor.Next()
[2022-05-09T16:30:59.877Z]       /home/jenkins/agent/workspace/ut-check/tidb/executor/executor.go:306 +0x645
[2022-05-09T16:30:59.877Z]   github.com/pingcap/tidb/executor.(*SelectionExec).Next()
[2022-05-09T16:30:59.877Z]       /home/jenkins/agent/workspace/ut-check/tidb/executor/executor.go:1398 +0x278
[2022-05-09T16:30:59.877Z]   github.com/pingcap/tidb/executor.Next()
[2022-05-09T16:30:59.877Z]       /home/jenkins/agent/workspace/ut-check/tidb/executor/executor.go:306 +0x645
[2022-05-09T16:30:59.877Z]   github.com/pingcap/tidb/executor.(*HashJoinExec).fetchBuildSideRows()
[2022-05-09T16:30:59.877Z]       /home/jenkins/agent/workspace/ut-check/tidb/executor/join.go:282 +0x43e
[2022-05-09T16:30:59.877Z]   github.com/pingcap/tidb/executor.(*HashJoinExec).fetchAndBuildHashTable.func2()
[2022-05-09T16:30:59.877Z]       /home/jenkins/agent/workspace/ut-check/tidb/executor/join.go:731 +0xf9
[2022-05-09T16:30:59.877Z]   github.com/pingcap/tidb/util.WithRecovery()
[2022-05-09T16:30:59.877Z]       /home/jenkins/agent/workspace/ut-check/tidb/util/misc.go:90 +0x6d
[2022-05-09T16:30:59.877Z]   github.com/pingcap/tidb/executor.(*HashJoinExec).fetchAndBuildHashTable.func4()
[2022-05-09T16:30:59.877Z]       /home/jenkins/agent/workspace/ut-check/tidb/executor/join.go:728 +0x47
[2022-05-09T16:30:59.877Z] 
[2022-05-09T16:30:59.877Z] Goroutine 68 (running) created at:
[2022-05-09T16:30:59.877Z]   github.com/pingcap/tidb/executor.(*HashJoinExec).fetchAndBuildHashTable()
[2022-05-09T16:30:59.877Z]       /home/jenkins/agent/workspace/ut-check/tidb/executor/join.go:728 +0x405
[2022-05-09T16:30:59.877Z]   github.com/pingcap/tidb/executor.(*HashJoinExec).Next.func1()
[2022-05-09T16:30:59.877Z]       /home/jenkins/agent/workspace/ut-check/tidb/executor/join.go:687 +0xc8
[2022-05-09T16:30:59.877Z]   github.com/pingcap/tidb/util.WithRecovery()
[2022-05-09T16:30:59.877Z]       /home/jenkins/agent/workspace/ut-check/tidb/util/misc.go:90 +0x6d
[2022-05-09T16:30:59.877Z]   github.com/pingcap/tidb/executor.(*HashJoinExec).Next.func2()
[2022-05-09T16:30:59.877Z]       /home/jenkins/agent/workspace/ut-check/tidb/executor/join.go:685 +0x47
```

```
[2022-05-09T16:30:59.883Z] ==================
[2022-05-09T16:30:59.883Z] WARNING: DATA RACE
[2022-05-09T16:30:59.883Z] Write at 0x00c0050eb9c0 by goroutine 68:
[2022-05-09T16:30:59.883Z]   github.com/pingcap/tidb/executor.tryFillViewColumnType.func1()
[2022-05-09T16:30:59.883Z]       /home/jenkins/agent/workspace/ut-check/tidb/executor/show.go:1913 +0x645
[2022-05-09T16:30:59.883Z]   github.com/pingcap/tidb/executor.runWithSystemSession()
[2022-05-09T16:30:59.883Z]       /home/jenkins/agent/workspace/ut-check/tidb/executor/show.go:1940 +0x334
[2022-05-09T16:30:59.883Z]   github.com/pingcap/tidb/executor.tryFillViewColumnType()
[2022-05-09T16:30:59.883Z]       /home/jenkins/agent/workspace/ut-check/tidb/executor/show.go:1904 +0x16d
[2022-05-09T16:30:59.883Z]   github.com/pingcap/tidb/executor.(*hugeMemTableRetriever).dataForColumnsInTable()
[2022-05-09T16:30:59.883Z]       /home/jenkins/agent/workspace/ut-check/tidb/executor/infoschema_reader.go:680 +0x159
[2022-05-09T16:30:59.883Z]   github.com/pingcap/tidb/executor.(*hugeMemTableRetriever).setDataForColumns()
[2022-05-09T16:30:59.883Z]       /home/jenkins/agent/workspace/ut-check/tidb/executor/infoschema_reader.go:669 +0x4a4
[2022-05-09T16:30:59.883Z]   github.com/pingcap/tidb/executor.(*hugeMemTableRetriever).retrieve()
[2022-05-09T16:30:59.883Z]       /home/jenkins/agent/workspace/ut-check/tidb/executor/infoschema_reader.go:2614 +0x314
[2022-05-09T16:30:59.883Z]   github.com/pingcap/tidb/executor.(*MemTableReaderExec).Next()
[2022-05-09T16:30:59.883Z]       /home/jenkins/agent/workspace/ut-check/tidb/executor/memtable_reader.go:118 +0x482
[2022-05-09T16:30:59.883Z]   github.com/pingcap/tidb/executor.Next()
[2022-05-09T16:30:59.883Z]       /home/jenkins/agent/workspace/ut-check/tidb/executor/executor.go:306 +0x645
[2022-05-09T16:30:59.883Z]   github.com/pingcap/tidb/executor.(*SelectionExec).Next()
[2022-05-09T16:30:59.883Z]       /home/jenkins/agent/workspace/ut-check/tidb/executor/executor.go:1398 +0x278
[2022-05-09T16:30:59.883Z]   github.com/pingcap/tidb/executor.Next()
[2022-05-09T16:30:59.883Z]       /home/jenkins/agent/workspace/ut-check/tidb/executor/executor.go:306 +0x645
[2022-05-09T16:30:59.883Z]   github.com/pingcap/tidb/executor.(*HashJoinExec).fetchBuildSideRows()
[2022-05-09T16:30:59.883Z]       /home/jenkins/agent/workspace/ut-check/tidb/executor/join.go:282 +0x43e
[2022-05-09T16:30:59.883Z]   github.com/pingcap/tidb/executor.(*HashJoinExec).fetchAndBuildHashTable.func2()
[2022-05-09T16:30:59.883Z]       /home/jenkins/agent/workspace/ut-check/tidb/executor/join.go:731 +0xf9
[2022-05-09T16:30:59.883Z]   github.com/pingcap/tidb/util.WithRecovery()
[2022-05-09T16:30:59.883Z]       /home/jenkins/agent/workspace/ut-check/tidb/util/misc.go:90 +0x6d
[2022-05-09T16:30:59.883Z]   github.com/pingcap/tidb/executor.(*HashJoinExec).fetchAndBuildHashTable.func4()
[2022-05-09T16:30:59.883Z]       /home/jenkins/agent/workspace/ut-check/tidb/executor/join.go:728 +0x47
[2022-05-09T16:30:59.883Z] 
[2022-05-09T16:30:59.883Z] Previous read at 0x00c0050eb9c0 by goroutine 66:
[2022-05-09T16:30:59.883Z]   github.com/pingcap/tidb/parser/types.(*FieldType).CompactStr()
[2022-05-09T16:30:59.883Z]       /home/jenkins/agent/workspace/ut-check/tidb/parser/types/field_type.go:220 +0x134
[2022-05-09T16:30:59.883Z]   github.com/pingcap/tidb/parser/model.(*ColumnInfo).GetTypeDesc()
[2022-05-09T16:30:59.883Z]       /home/jenkins/agent/workspace/ut-check/tidb/parser/model/model.go:271 +0x39
[2022-05-09T16:30:59.883Z]   github.com/pingcap/tidb/table.NewColDesc()
[2022-05-09T16:30:59.883Z]       /home/jenkins/agent/workspace/ut-check/tidb/table/column.go:404 +0x704
[2022-05-09T16:30:59.883Z]   github.com/pingcap/tidb/executor.(*hugeMemTableRetriever).dataForColumnsInTable()
[2022-05-09T16:30:59.883Z]       /home/jenkins/agent/workspace/ut-check/tidb/executor/infoschema_reader.go:793 +0x1044
[2022-05-09T16:30:59.883Z]   github.com/pingcap/tidb/executor.(*hugeMemTableRetriever).setDataForColumns()
[2022-05-09T16:30:59.883Z]       /home/jenkins/agent/workspace/ut-check/tidb/executor/infoschema_reader.go:669 +0x4a4
[2022-05-09T16:30:59.883Z]   github.com/pingcap/tidb/executor.(*hugeMemTableRetriever).retrieve()
[2022-05-09T16:30:59.883Z]       /home/jenkins/agent/workspace/ut-check/tidb/executor/infoschema_reader.go:2614 +0x314
[2022-05-09T16:30:59.883Z]   github.com/pingcap/tidb/executor.(*MemTableReaderExec).Next()
[2022-05-09T16:30:59.883Z]       /home/jenkins/agent/workspace/ut-check/tidb/executor/memtable_reader.go:118 +0x482
[2022-05-09T16:30:59.883Z]   github.com/pingcap/tidb/executor.Next()
[2022-05-09T16:30:59.883Z]       /home/jenkins/agent/workspace/ut-check/tidb/executor/executor.go:306 +0x645
[2022-05-09T16:30:59.883Z]   github.com/pingcap/tidb/executor.(*SelectionExec).Next()
[2022-05-09T16:30:59.883Z]       /home/jenkins/agent/workspace/ut-check/tidb/executor/executor.go:1398 +0x278
[2022-05-09T16:30:59.883Z]   github.com/pingcap/tidb/executor.Next()
[2022-05-09T16:30:59.883Z]       /home/jenkins/agent/workspace/ut-check/tidb/executor/executor.go:306 +0x645
[2022-05-09T16:30:59.883Z]   github.com/pingcap/tidb/executor.(*HashJoinExec).fetchBuildSideRows()
[2022-05-09T16:30:59.883Z]       /home/jenkins/agent/workspace/ut-check/tidb/executor/join.go:282 +0x43e
[2022-05-09T16:30:59.883Z]   github.com/pingcap/tidb/executor.(*HashJoinExec).fetchAndBuildHashTable.func2()
[2022-05-09T16:30:59.883Z]       /home/jenkins/agent/workspace/ut-check/tidb/executor/join.go:731 +0xf9
[2022-05-09T16:30:59.883Z]   github.com/pingcap/tidb/util.WithRecovery()
[2022-05-09T16:30:59.883Z]       /home/jenkins/agent/workspace/ut-check/tidb/util/misc.go:90 +0x6d
[2022-05-09T16:30:59.883Z]   github.com/pingcap/tidb/executor.(*HashJoinExec).fetchAndBuildHashTable.func4()
[2022-05-09T16:30:59.883Z]       /home/jenkins/agent/workspace/ut-check/tidb/executor/join.go:728 +0x47
[2022-05-09T16:30:59.883Z] 
[2022-05-09T16:30:59.883Z] Goroutine 68 (running) created at:
[2022-05-09T16:30:59.883Z]   github.com/pingcap/tidb/executor.(*HashJoinExec).fetchAndBuildHashTable()
[2022-05-09T16:30:59.883Z]       /home/jenkins/agent/workspace/ut-check/tidb/executor/join.go:728 +0x405
[2022-05-09T16:30:59.883Z]   github.com/pingcap/tidb/executor.(*HashJoinExec).Next.func1()
[2022-05-09T16:30:59.883Z]       /home/jenkins/agent/workspace/ut-check/tidb/executor/join.go:687 +0xc8
[2022-05-09T16:30:59.883Z]   github.com/pingcap/tidb/util.WithRecovery()
[2022-05-09T16:30:59.883Z]       /home/jenkins/agent/workspace/ut-check/tidb/util/misc.go:90 +0x6d
[2022-05-09T16:30:59.883Z]   github.com/pingcap/tidb/executor.(*HashJoinExec).Next.func2()
[2022-05-09T16:30:59.883Z]       /home/jenkins/agent/workspace/ut-check/tidb/executor/join.go:685 +0x47
```

### 4. What is your TiDB version? (Required)

It's introduced by https://github.com/pingcap/tidb/pull/33946/files#diff-e73beb95453376eb481861be1aeba193a26801a5fb13b20e04dade9f24c5d128R1937

### PR DESCRIPTION
**Pull request by tiancaiamao**
### What problem does this PR solve?

Issue Number: close #34509

Problem Summary:

When reading from `information_schema.columns`, the data for `view` are special.
Because unlike a normal table, the columns of a `view` is generated from other plan.  
The old code use `tryFillViewColumnType()` function to **modify** the table column information.

Note, this table column information (infoschema tableInfo) is shared by all the whole process,
it's designed to be read-only (use atomic replacement in `domain.Reload()` as a whole).
Modify this is not thread-safe, especially in `join` executor, there're many join worker process. 

### What is changed and how it works?

Avoid using `tryFillViewColumnType` which modify the column of tableInfo.Column.

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)

Check the DATA RACE is fixed:

make ut X='run executor --race'

- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

**Comment by tiancaiamao** [comment]
The original PR for handle view data is https://github.com/pingcap/tidb/pull/22363, IMHO it's not correct.
It notice the data race problem and commented [here](https://github.com/pingcap/tidb/pull/22363/files#diff-e73beb95453376eb481861be1aeba193a26801a5fb13b20e04dade9f24c5d128L463-L464)

The later fix https://github.com/pingcap/tidb/pull/33946/ try to use `sysSession` to avoid the data race is ... also a wrong direction.

File: executor/infoschema_reader.go [hunk 1]
```diff
@@ -677,10 +678,20 @@ func (e *hugeMemTableRetriever) setDataForColumns(ctx context.Context, sctx sess
 }
 
 func (e *hugeMemTableRetriever) dataForColumnsInTable(ctx context.Context, sctx sessionctx.Context, schema *model.DBInfo, tbl *model.TableInfo, priv mysql.PrivilegeType, extractor *plannercore.ColumnsTableExtractor) {
-	if err := tryFillViewColumnType(ctx, sctx, sctx.GetInfoSchema().(infoschema.InfoSchema), schema.Name, tbl); err != nil {
-		sctx.GetSessionVars().StmtCtx.AppendWarning(err)
-		return
+	is := sctx.GetInfoSchema().(infoschema.InfoSchema)
```
Comment (hawkingrei) [inline executor/infoschema_reader.go:681]: ```GetInfoSchema``` is Deprecated. we can use ```sessiontxn.GetTxnManager(ctx).GetTxnInfoSchema()``` to replace.

Comment (tiancaiamao) [reply executor/infoschema_reader.go]: Done @hawkingrei

**Review by hawkingrei — APPROVED** [review]

**Review by xhebox — APPROVED** [review]

**Pull request merged**